Fix null reference handling in VSD stub for x86

janvorli opened the pull request:
The regression test that I've added recently to accompany a fix for null
reference handling in VSD stub for x64 is failing for x86. The problem
is that the null reference handling in VSD dispatch and resolve stubs
was broken in another way due to the x86 calling convention. When the
call went through a shuffle thunk that removes one stack argument due to
the shuffle, the manual unwinding in AdjustContextForVirtualStub was
getting an ESP that was off by one stack slot and exception handling
wasn't able to correctly unwind from that location to the caller.

This change fixes it by letting the AdjustContextForVirtualStub manually
unwind to the instruction after the call to the shuffle thunk / VSD stub
and updating the ESP according to the number of stack arguments of the
target method.

I have also modified the regression test. One change was to make a call
with multiple parameters to verify that the logic to get stack arguments
size is working correctly. Another change was to make sure that both
dispatch and resolve stub cases are tested.

Close #35188

janvorli committed:
Fix null reference handling in VSD stub for x86

The regression test that I've added recently to accompany a fix for null
reference handling in VSD stub for x64 is failing for x86. The problem
is that the null reference handling in VSD dispatch and resolve stubs
was broken in another way due to the x86 calling convention. When the
call went through a shuffle thunk that removes one stack argument due to
the shuffle, the manual unwinding in AdjustContextForVirtualStub was
getting an ESP that was off by one stack slot and exception handling
wasn't able to correctly unwind from that location to the caller.

This change fixes it by letting the AdjustContextForVirtualStub manually
unwind to the instruction after the call to the shuffle thunk / VSD stub
and updating the ESP according to the number of stack arguments of the
target method.

I have also modified the regression test. One change was to make a call
with multiple parameters to verify that the logic to get stack arguments
size is working correctly. Another change was to make sure that both
dispatch and resolve stub cases are tested.

janvorli commented:
/azp run runtime-coreclr outerloop

jkotas commented on src/coreclr/src/vm/virtualcallstub.cpp:
```diff
@@ -1914,7 +1914,11 @@ PCODE VirtualCallStubManager::ResolveWorker(StubCallSite* pCallSite,
-                                                             token.To_SIZE_T());
+                                                             token.To_SIZE_T()
+#ifdef TARGET_X86
+                                                             , objectType
```
Does the `objectType` have the right lifetime? Can the resolve stub outlive it?

(Look for `bCallToShorterLivedTarget` above for some context.)

janvorli replied:
Thank you for pointing it out, it seems it can happen. It seems that I can fix that by storing the stack arguments size in the stub instead of the MethodTable for the resolve stub.

janvorli replied:
I've just pushed a commit that replaces the usage of MethodTable by the size of stack arguments in the ResolveStub.

janvorli committed:
Replace MethodTable in the ResolveStub by size of stack args

This prevents issues in case the type represented by the MethodTable got
unloaded.

jkotas commented on src/coreclr/src/vm/i386/virtualcallstubcpu.hpp:
```diff
@@ -329,6 +331,7 @@ struct ResolveStub
     DISPL _backpatcherDispl;        // xx xx xx xx          backpatcherWorker  == BackPatchWorkerAsmStub
     BYTE  part11 [1];               // eb           jmp
     BYTE toResolveStub;             // xx                   resolveStub, i.e. go back to _resolveEntryPoint
+    size_t _stackArgumentsSize;     // xx xx xx xx
```
Should all this be under `#ifndef UNIX_X86_ABI` ?

janvorli replied:
You are right, I've forgotten that on x86 Linux we use caller side stack arguments cleanup.

janvorli committed:
Make the stack arguments size stuff Windows specific

On Unix x86, the stack is cleaned up by the caller, not the callee.
